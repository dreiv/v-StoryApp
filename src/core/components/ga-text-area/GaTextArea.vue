<script setup lang="ts">
import {
  computed,
  inject,
  useAttrs,
  useCssModule,
  useTemplateRef,
  type TextareaHTMLAttributes,
} from 'vue'
import { formFieldKey } from '../ga-form-field/types'

export interface GaTextAreaProps extends /* @vue-ignore */ TextareaHTMLAttributes {
  id?: string
  error?: boolean
  success?: boolean
}

const { id, error, disabled, success } = defineProps<GaTextAreaProps>()

const inputRef = useTemplateRef<HTMLTextAreaElement>('inputRef')
const style = useCssModule()
const attrs = useAttrs()
const model = defineModel<string>()

const formField = inject(formFieldKey)

const inputAttrs = computed(() => ({
  ...attrs,
  id: id || formField?.id,
  disabled: disabled || formField?.disabled,
}))

const classes = computed(() => [
  style.text,
  { [style.error]: error || formField?.error, [style.success]: success },
])

defineExpose({ inputRef })
</script>

<template>
  <textarea ref="inputRef" rows="5" :class="classes" v-model="model" v-bind="inputAttrs" />
</template>

<style module>
.text {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-primary);
  border-radius: var(--ga-radius);
  background-color: var(--ga-color-surface-primary);

  padding: var(--ga-size-spacing-03) var(--ga-size-spacing-04);
  width: 100%;
  min-width: 0rem;

  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-md-tracking);

  &:hover {
    border-color: var(--ga-color-border-action-hover);
  }

  &:disabled {
    border-color: var(--ga-color-border-primary);
    background-color: var(--ga-color-surface-disabled);
    color: var(--ga-color-text-disable-selected);
  }

  &:focus:not(.error, .success) {
    outline: 1px solid var(--ga-color-border-focus);
    border-color: var(--ga-color-border-focus);
  }

  &.error {
    outline: 1px solid var(--ga-color-border-error);
    border-color: var(--ga-color-border-error);
    background-color: var(--ga-color-surface-error);
  }
}
</style>
