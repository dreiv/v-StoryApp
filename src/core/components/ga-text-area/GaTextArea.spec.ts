import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaTextArea from './GaTextArea.vue'

describe('GaTextArea', () => {
  let wrapper: VueWrapper

  it('should render correctly', () => {
    wrapper = shallowMount(GaTextArea)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply $className class when $propName prop is true', () => {
    wrapper = shallowMount(GaTextArea, { props: { error: true } })
    expect(wrapper.find('textarea').classes()).toContain('error')
  })

  it('should bind value with v-model', async () => {
    wrapper = shallowMount(GaTextArea, {
      props: { modelValue: 'initial text' },
    })

    const textarea = wrapper.find('textarea')
    await textarea.setValue('updated text')

    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual(['updated text'])
  })

  it('should set disabled attribute when disabled prop is true', async () => {
    wrapper = shallowMount(GaTextArea, {
      props: { disabled: true },
    })

    const textarea = wrapper.find('textarea')
    expect(textarea.element.disabled).toBe(true)
  })

  it('should pass attributes to the textarea element', () => {
    wrapper = shallowMount(GaTextArea, {
      attrs: {
        placeholder: 'Enter description',
        maxlength: '100',
      },
    })

    const textarea = wrapper.find('textarea')
    expect(textarea.attributes('placeholder')).toBe('Enter description')
    expect(textarea.attributes('maxlength')).toBe('100')
  })
})
