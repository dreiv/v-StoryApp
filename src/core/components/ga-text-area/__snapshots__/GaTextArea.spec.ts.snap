// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaTextArea > should render correctly 1`] = `"<textarea rows="5" class="text"></textarea>"`;
